<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [16]:
# Параметры для подключения берем из файла config.txt 
params = {k: v.strip() for k, v in (l.split('=') for l in open('config.txt'))}


In [17]:
connection = psycopg2.connect(
    dbname=params['DBNAME'],
    user=params['USER'],
    host=params['HOST'],
    password=params['PASSWORD'],
    port=int(params['PORT'])
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [18]:
# текст запроса
query_3_1 = f'''select count(*)
from vacancies
'''

In [19]:
# результат запроса
vacancies = pd.read_sql_query(query_3_1, connection)
vacancies


C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\2081210222.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [20]:
# текст запроса
query_3_2 = f'''select count(*)
from employers
'''

In [21]:
# результат запроса
employers = pd.read_sql_query(query_3_2, connection)
employers

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\1389401573.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [22]:
# текст запроса
query_3_3 = f'''select count(*)
from areas
'''

In [24]:
# результат запроса
areas = pd.read_sql_query(query_3_3, connection)
areas

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\744816049.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [25]:
# текст запроса
query_3_4 = f'''select count(*)
from industries
'''

In [26]:
# результат запроса
industries = pd.read_sql_query(query_3_4, connection)
industries

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\2016089628.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных

### Выводы
1. В таблице vacancies представленно 49197 записей (строк) о вакансиях. Столбцы key_skills, salary_from и salary_to имеют пропуски. Количество уникальных названий вакансий (столбец name) составляет 21223.
2. В таблице employers представленно 23501 записей о работадателях. Количество уникальных названий работадателей (столбец name) составляет 23175.
3. В таблице araes представленно 1362 записей о регионах.
4. В таблице industries представленно 294 записей о сферах деятельности работадателей.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [27]:
# текст запроса
query_4_1 = f'''select 
    a.name area,
    count(v.id) cnt
from vacancies v
join areas a on a.id = v.area_id
group by a.id
order by 2 desc
limit 5
'''

In [28]:
# результат запроса
cnt_vac_area = pd.read_sql_query(query_4_1, connection)
cnt_vac_area

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\4289463329.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac_area = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [29]:
# текст запроса
query_4_2 = f'''select 
    count(id) cnt
from vacancies
where salary_from is not null or salary_to is not null
'''

In [30]:
# результат запроса
cnt_vac = pd.read_sql_query(query_4_2, connection)
cnt_vac

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\2300707370.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [31]:
# текст запроса
query_4_3 = f'''select 
    round(avg(salary_from)) low_bound,
    round(avg(salary_to)) upp_bound
from vacancies
'''

In [32]:
# результат запроса
avg_sal = pd.read_sql_query(query_4_3, connection)
avg_sal

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\16086132.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_sal = pd.read_sql_query(query_4_3, connection)


low_bound  upp_bound
0    71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [33]:
# текст запроса
query_4_4 = f'''select 
    schedule,
    employment,
    count(id) cnt
from vacancies
group by schedule, employment
order by 3 desc
'''

In [34]:
# результат запроса
cnt_vac_sched_emp = pd.read_sql_query(query_4_4, connection)
cnt_vac_sched_emp

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\422935498.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac_sched_emp = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [35]:
# текст запроса
query_4_5 = f'''select 
    experience,
    count(id) cnt
from vacancies
group by experience
order by 2
'''

In [36]:
# результат запроса
cnt_vac_exp = pd.read_sql_query(query_4_5, connection)
cnt_vac_exp

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\529240280.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac_exp = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [ ]:
# выводы по детальному анализу вакансий

### Выводы
1. Топ пять регионов по количеству вакансий (по убыванию) следующий: Москва,Санкт-Петербург, Минск, Новосибирск, Алматы
2. У 24073 записей в таблице с вакансиями заполнено хотя бы одно из полей с зарплатой, что составляет менее 50 %.
3. Среднее значение нижней границы зарплатной вилки составляет 71065 руб, верхней границы - 110537 руб.
4. Второе по численности вакансий сочетание типа рабочего графика (schedule) и типа трудоустройства (employment) является - "Удалённая работа — Полная занятость"
5. Требуемый опыт работы по количеству вакансий (в порядке возрастания) представляет собой следующую последовательность: Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [37]:
# текст запроса
query_5_1 = f'''select 
    e.name,
    count(v.id) cnt
from vacancies v
join employers e on e.id = v.employer_id
group by e.name
order by 2 desc
limit 5
'''

In [38]:
# результат запроса
cnt_vac_emp = pd.read_sql_query(query_5_1, connection)
cnt_vac_emp

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\3981223690.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac_emp = pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [39]:
# текст запроса
query_5_2 = f'''select 
    a.name,
    count(e.id) cnt_emp,
    count(v.id) cnt_vac
from areas a
left join vacancies v on a.id = v.area_id
left join employers e on a.id = e.area
group by a.name
order by 3, 2 desc
'''
# При подсчете количества уникальных значений id работадателя с использованием distinct время обработки запроса слишком длительное (результата я не дождался).
# Аналогичная ситуация и при группировке по id работадателя

In [40]:
# результат запроса
area_vac_emp = pd.read_sql_query(query_5_2, connection)
area_vac_emp.head()

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\270858729.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  area_vac_emp = pd.read_sql_query(query_5_2, connection)


name  cnt_emp  cnt_vac
0              Россия      410        0
1           Казахстан      207        0
2  Московская область       75        0
3  Краснодарский край       19        0
4  Ростовская область       18        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [41]:
# текст запроса
query_5_3 = f'''select 
    e.name,
    count(distinct a.id) cnt_area
from vacancies v
join areas a on v.area_id = a.id
join employers e on e.id = v.employer_id
group by e.name
order by 2 desc
limit 5
'''

In [42]:
# результат запроса
emp_cnt_area = pd.read_sql_query(query_5_3, connection)
emp_cnt_area

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\582158040.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_cnt_area = pd.read_sql_query(query_5_3, connection)


name  cnt_area
0                  Яндекс       181
1              Ростелеком       152
2              Спецремонт       116
3  Поляков Денис Иванович        88
4               ООО ЕФИН         71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [12]:
# текст запроса
query_5_4 = f'''select 
    count(e.id)
from employers e
left join employers_industries ei on e.id = ei.employer_id
where ei.industry_id is null
'''

In [13]:
# результат запроса
emp_cny_null_ind = pd.read_sql_query(query_5_4, connection)
emp_cny_null_ind

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_11012\337339440.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_cny_null_ind = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [16]:
# текст запроса
query_5_5 = f'''select 
    e.name
from employers e
full outer join employers_industries ei on e.id = ei.employer_id
group by e.name
having count(ei.industry_id) = 4
order by 1
offset 2 limit 1
'''

In [17]:
# результат запроса
third_emp = pd.read_sql_query(query_5_5, connection)
third_emp

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_11012\1556874130.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  third_emp = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [18]:
# текст запроса
query_5_6 = f'''select 
    count(e.id) cnt
from employers_industries ei
join employers e on e.id = ei.employer_id
join industries i on i.id = ei.industry_id
group by i.name
having i.name = 'Разработка программного обеспечения'
'''

In [19]:
# результат запроса
cnt_emp_ind = pd.read_sql_query(query_5_6, connection)
cnt_emp_ind

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_11012\704479369.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_emp_ind = pd.read_sql_query(query_5_6, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
table = pd.read_html(url)
# Оформляем список городов-миллионников в виде кортежа
city_mil = tuple(table[0]['Город'])

In [49]:
# текст запроса
query_5_7 = f'''(select 
    a.name "Город",
    count(v.employer_id) cnt
from vacancies v
join employers e on e.id = v.employer_id
join areas a on a.id = v.area_id
group by a.name, e.name
having e.name = 'Яндекс' and a.name in {city_mil})
union
 (select 
    'Total',
    count(*)
from vacancies v
join employers e on e.id = v.employer_id
join areas a on a.id = v.area_id
where a.name in {city_mil}
group by e.name
having e.name = 'Яндекс')
order by 2 
'''

In [ ]:
# результат запроса
table_city_yandex = pd.read_sql_query(query_5_7, connection)
table_city_yandex

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\2632619327.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  table_city_yandex = pd.read_sql_query(query_5_7, connection)


Город  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

In [ ]:
# выводы по анализу работодателей

### Выводы
1. Топ пять работадателей по количеству вакансий возлавляет Яндекс, а замыкает Газпром нефть.
2. Регион с наибольшим количеством работадателей, при отсутствии вакансий (очевидно, по причине отсутствия указания именно такого названия региона) является Россия.
3. Работадатель с наибольшим количеством регионов, где он размещает свои вакансии - Яндекс, количество регионов - 181.
4. У 8419 работадателей не указана сфера деятельности.
5. Третья по алфавиту компания с четырьмя сферами деятельности - 2ГИС.
6. В качестве сферы деятельности указана «Разработка программного обеспечения» у 3553 работадателей.
7. Яндекс представляет вакансии во всех 16 городах-миллионниках России, с общим колчеством вакансий по этим городам - 485.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [51]:
# текст запроса
query_6_1 = f'''select 
    count(id) cnt
from vacancies
where lower(name) like '%data%' or lower(name) like '%данн%'
'''

In [52]:
# результат запроса
cnt_vac_data = pd.read_sql_query(query_6_1, connection)
cnt_vac_data

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\2511655081.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac_data = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [65]:
# текст запроса
query_6_2 = f'''select 
    count(id) cnt
from vacancies
where ((lower(name) like '%data scientist%') or (lower(name) like '%data science%') 
or (lower(name) like '%исследователь данных%') or (name like '%ML%' and name not ilike '%HTML%' and name not ilike '%QML%')
or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) 
and (lower(name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка')
'''

In [66]:
# результат запроса
cnt_jun_data = pd.read_sql_query(query_6_2, connection)
cnt_jun_data

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_10824\12764407.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_jun_data = pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [73]:
# текст запроса
query_6_3 = f'''select 
    count(id) cnt
from vacancies
where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' or (name ilike '%ML%' and name not ilike '%HTML%')
or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%') 
and (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
'''

In [74]:
# результат запроса
cnt_data_sql = pd.read_sql_query(query_6_3, connection)
cnt_data_sql

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_10824\3846396167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_data_sql = pd.read_sql_query(query_6_3, connection)


cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [55]:
# текст запроса
query_6_4 = f'''select 
    count(id) cnt
from vacancies v
where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' or (name ilike '%ML%' and upper(name) not like '%HTML%')
or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%') 
and (key_skills ilike '%python%')
'''

In [56]:
# результат запроса
cnt_data_py = pd.read_sql_query(query_6_4, connection)
cnt_data_py

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_10824\2890614297.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_data_py = pd.read_sql_query(query_6_4, connection)


cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [63]:
# текст запроса
query_6_5 = f'''select 
    round(avg(length(key_skills) - length(replace(key_skills,CHR(9),'')) + 1), 2)
from vacancies
where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' or ((name like '%ML%') and (name not ilike '%HTML%') and (name not ilike '%XML%'))
or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
and (key_skills is not null)
'''

In [64]:
# результат запроса
cnt_key_skills = pd.read_sql_query(query_6_5, connection)
cnt_key_skills

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_10824\3676575775.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_key_skills = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [4]:
# текст запроса
query_6_6 = f'''select 
    experience,
    round(avg(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)))
from vacancies
where lower(name) like '%data scientist%' or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not ilike '%HTML%')
or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%'
group by experience
'''

In [5]:
# результат запроса
avg_sal_ds = pd.read_sql_query(query_6_6, connection)
avg_sal_ds

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_12560\2333767249.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_sal_ds = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

In [ ]:
# выводы по предметному анализу

### Выводы
1. Среди всех вакансий 1771 имеют отношение к данным, в её названии содержатся слова 'data' или 'данн'.
2. Количество вакансий для начинающего дата-сайентиста (DS) составляет 51.
3. Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres, составляет 229.
4. Количество вакансий для DS, в которых в качестве ключевого навыка указан Python, составляет 357.
5. В среднем в вакансиях для DS указывают 6.41 ключевых навыков.
6. В среднем в вакансиях для DS с опытом работы от 3 до 6 лет указывают зарплату в 243115 руб.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Выведем распределение количества вакансий по сферам деятельности и оставим только те сферы деятельности, у которых количество вакансий больше или равно 1000.

In [ ]:
# текст запроса
query_add_1 = f'''select 
    i.name,
    count(*)
from vacancies v
join employers e on e.id = v.employer_id
join employers_industries ei on ei.employer_id = e.id
join industries i on i.id = ei.industry_id
group by i.name
having count(*) >= 1000
order by 2 desc
'''

In [ ]:
# результат запроса
add_table1 = pd.read_sql_query(query_add_1, connection)
add_table1

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\1871032506.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  add_table1 = pd.read_sql_query(query_add_1, connection)


name  count
0                 Разработка программного обеспечения  12499
1   Системная интеграция,  автоматизации технологи...  11034
2   Интернет-компания (поисковики, платежные систе...   6413
3                                                Банк   2742
4                                     Мобильная связь   1585
5                                  Кадровые агентства   1143
6                               Консалтинговые услуги   1137
7                                  Интернет-провайдер   1095
8                        Розничная сеть (продуктовая)   1034
9                                 Фиксированная связь   1028
10                               Оптоволоконная связь   1012
11  Маркетинговые, рекламные, BTL, дизайнерские, E...   1001

Выведем десять самых нуждающихся сфер деятельности, где требуется DS

In [ ]:
# текст запроса
query_add_2 = f'''select 
    i.name,
    count(*)
from vacancies v
join employers e on e.id = v.employer_id
join employers_industries ei on ei.employer_id = e.id
join industries i on i.id = ei.industry_id
where (lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' 
or lower(v.name) like '%исследователь данных%' or ((v.name like '%ML%') and (v.name not ilike '%HTML%') and (v.name not ilike '%XML%'))
or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%')
group by i.name
order by 2 desc
limit 10
'''

In [ ]:
# результат запроса
add_table2 = pd.read_sql_query(query_add_2, connection)
add_table2

C:\Users\Kurumir\AppData\Local\Temp\ipykernel_13608\1495963960.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  add_table2 = pd.read_sql_query(query_add_2, connection)


name  count
0                Разработка программного обеспечения    148
1  Системная интеграция,  автоматизации технологи...    139
2                                               Банк     92
3  Интернет-компания (поисковики, платежные систе...     79
4                                    Мобильная связь     23
5  Маркетинговые, рекламные, BTL, дизайнерские, E...     22
6  Управляющая, инвестиционная компания (управлен...     21
7                              Консалтинговые услуги     16
8                                Фиксированная связь     13
9  Научно-исследовательская, научная, академическ...      9

### Выводы по дополнительным исследованиям:
1. Тройка самых распространенных сфер деятельности - "Разработка программного обеспечения" (12499 вакансии), "Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг" (11034 вакансии), "Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее)" (6413 вакансии)
2. Дата-саентист наиболее востребован в следующих трёх сферах деятельности: "Разработка программного обеспечения" (148 вакансий), "Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг" (139 вакансий) и "Банк" (92 вакансии).

### Выводы:
1. Данные представлены в 5 таблицах. Основная таблица - vacancies содержит 49197 записей о вакансиях. Три таблицы-справочника - areas (1392 записи), employers (23501 запись) и industries (294 записи), содержащие информацию о регионе, работадателе и сфере деятельности соответственно. И одна дополнительная таблица - employers_industries для организации связи между работодателями и сферами их деятельности.
2. В ходе детального анализа данных выявили пятерку городов по количеству вакансий - Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. 
3. В таблице vacancies имеется большое количество пропусков в столбцах salary_from и salary_to - только у 24073 вакансий заполнено хотябы одно из этих полей. И среднее значение в них составляет 71065 руб для нижней границы (salary_from) и 110537 руб для верхней границы (salary_to)
4. Определили распределение количества вакансий по сочетанию типа рабочего графика (schedule) и типа трудоустройства (employment), а также распределение количества вакансий по требуемому опыту работы.
5. Определили распределение количества вакансий по работадателю, региону.
6. Опредлели распределение работадателей по регионам.
7. Рассмотрели сферы деятельности работадателей: у 8419 работадателей сфера деятельности не указана, а у 3553 работаделей в качестве сферы деятельности указана «Разработка программного обеспечения».
8. Рассмотрели распределение количества вакансий компании Яндекс в городах миллионниках России. Из 16 городов-миллионников, помимо Москвы с 54 вакансиями (11% от общего количества вакансий в городах-миллиониках) и Санкт-Петербурга с 42 вакансиями (> 9%), следуюет выделить города где количество вакансий 30 или больше : Краснодар (30), Воронеж (32), Новосибирск (35), Нижний Новгород (36) и Екатеринбург (39). В остальных городах количество вакансий распределено довольно равномерно и в среднем составляет примерно 24 вакансии.
9. Из всех представленных вакансий 1771 имеет отношение к данным. 
10. В ходе предметного анализа провели исследование на предмет количества вакансий, требований к ключевым навыкам, а также определили среднее значение зарплаты для дата-саентиста.
11. Дополнительно было рассмотрено распределение вакансий по сферам деятельности, в том числе рассмотрено в каких сферах наиболее востребована вакансия дата-саентиста.